File: apps/mac/Sources/ClipPlusMac/Settings/SettingsStore.swift
import Foundation

struct StoredSettings: Equatable {
    let sharedKeyConfigured: Bool
    let sharingEnabled: Bool
    let sharedGroupId: String
    let sharedKeyInput: String
}

final class SettingsStore {
    private enum Key {
        static let sharedGroupId = "clipplus.shared_group_id"
        static let sharingEnabled = "clipplus.sharing_enabled"
    }

    private let userDefaults: UserDefaults
    private let sharedKeyVault: SharedKeyVault

    init(
        userDefaults: UserDefaults = .standard,
        sharedKeyVault: SharedKeyVault = FileSharedKeyVault()
    ) {
        self.userDefaults = userDefaults
        self.sharedKeyVault = sharedKeyVault
    }

    func load() -> StoredSettings {
        let sharedGroupId = userDefaults.string(forKey: Key.sharedGroupId) ?? ""
        let sharingEnabled = userDefaults.object(forKey: Key.sharingEnabled) as? Bool ?? true
        let sharedKeyInput = sharedKeyVault.loadSharedKey()

        return StoredSettings(
            sharedKeyConfigured: !sharedGroupId.isEmpty && !sharedKeyInput.isEmpty,
            sharingEnabled: sharingEnabled,
            sharedGroupId: sharedGroupId,
            sharedKeyInput: sharedKeyInput
        )
    }

    func saveSharedGroupId(_ sharedGroupId: String) {
        let trimmedGroupId = sharedGroupId.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmedGroupId.isEmpty {
            userDefaults.removeObject(forKey: Key.sharedGroupId)
        } else {
            userDefaults.set(trimmedGroupId, forKey: Key.sharedGroupId)
        }
    }

    func saveSharedKey(_ sharedKey: String, sharedGroupId: String) throws {
        try sharedKeyVault.saveSharedKey(sharedKey)
        saveSharedGroupId(sharedGroupId)
    }

    func saveSharingEnabled(_ sharingEnabled: Bool) {
        userDefaults.set(sharingEnabled, forKey: Key.sharingEnabled)
    }
}

File: apps/windows/ClipPlus.Windows/Settings/SettingsStore.cs
using System.IO;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ClipPlus.Windows.Settings;

public sealed record PersistedSettings(
    bool SharedKeyConfigured,
    bool SharingEnabled,
    string SharedGroupId
)
{
    [JsonIgnore]
    public string SharedKeyInput { get; init; } = string.Empty;
}

public sealed class SettingsStore
{
    private const string FileName = "settings.json";
    private readonly string directory;
    private readonly string filePath;
    private readonly SharedKeyVault sharedKeyVault;

    public SettingsStore()
        : this(Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "ClipPlus"),
            new FileSharedKeyVault())
    {
    }

    public SettingsStore(string directory)
        : this(directory, new FileSharedKeyVault(Path.Combine(directory, FileSharedKeyVault.FileName)))
    {
    }

    public SettingsStore(string directory, SharedKeyVault sharedKeyVault)
    {
        this.directory = directory;
        filePath = Path.Combine(directory, FileName);
        this.sharedKeyVault = sharedKeyVault;
    }

    public PersistedSettings Load()
    {
        if (!File.Exists(filePath))
        {
            return DefaultSettings();
        }

        try
        {
            var settings = JsonSerializer.Deserialize<PersistedSettings>(File.ReadAllText(filePath))
                ?? DefaultSettings();
            var sharedGroupId = settings.SharedGroupId.Trim();
            var sharedKeyInput = sharedKeyVault.LoadSharedKey().Trim();

            return settings with
            {
                SharedKeyConfigured = settings.SharedKeyConfigured
                    && !string.IsNullOrEmpty(sharedGroupId)
                    && !string.IsNullOrEmpty(sharedKeyInput),
                SharedGroupId = sharedGroupId,
                SharedKeyInput = sharedKeyInput
            };
        }
        catch (JsonException)
        {
            return DefaultSettings();
        }
        catch (IOException)
        {
            return DefaultSettings();
        }
    }

    public void Save(SettingsState state)
    {
        Save(new PersistedSettings(
            SharedKeyConfigured: state.SharedKeyConfigured,
            SharingEnabled: state.SharingEnabled,
            SharedGroupId: state.SharedGroupId
        )
        {
            SharedKeyInput = state.SharedKeyInput
        });
    }

    public void Save(PersistedSettings settings)
    {
        Directory.CreateDirectory(directory);
        var sharedGroupId = settings.SharedGroupId.Trim();
        var sharedKeyInput = settings.SharedKeyInput.Trim();
        var normalizedSettings = settings with
        {
            SharedKeyConfigured = settings.SharedKeyConfigured
                && !string.IsNullOrEmpty(sharedGroupId)
                && !string.IsNullOrEmpty(sharedKeyInput),
            SharedGroupId = sharedGroupId,
            SharedKeyInput = sharedKeyInput
        };

        if (!string.IsNullOrEmpty(sharedKeyInput))
        {
            sharedKeyVault.SaveSharedKey(sharedKeyInput);
        }

        File.WriteAllText(filePath, JsonSerializer.Serialize(normalizedSettings));
    }

    private static PersistedSettings DefaultSettings()
    {
        return new PersistedSettings(
            SharedKeyConfigured: false,
            SharingEnabled: true,
            SharedGroupId: string.Empty
        );
    }
}
